File: library/src/main/java/com/afollestad/aesthetic/AestheticImageView.java
package com.afollestad.aesthetic;

import static com.afollestad.aesthetic.Rx.onErrorLogAndRethrow;
import static com.afollestad.aesthetic.Util.resolveResId;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatImageView;
import android.util.AttributeSet;
import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;

/** @author Aidan Follestad (afollestad) */
public class AestheticImageView extends AppCompatImageView {

  private Disposable bgSubscription;
  private int backgroundResId;

  public AestheticImageView(Context context) {
    super(context);
  }

  public AestheticImageView(Context context, @Nullable AttributeSet attrs) {
    super(context, attrs);
    init(context, attrs);
  }

  public AestheticImageView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
    init(context, attrs);
  }

  private void init(Context context, AttributeSet attrs) {
    if (attrs != null) {
      backgroundResId = resolveResId(context, attrs, android.R.attr.background);
    }
  }

  @Override
  protected void onAttachedToWindow() {
    super.onAttachedToWindow();
    Observable<Integer> obs = ViewUtil.getObservableForResId(getContext(), backgroundResId, null);
    if (obs != null) {
      bgSubscription =
          obs.compose(Rx.<Integer>distinctToMainThread())
              .subscribe(ViewBackgroundAction.create(this), onErrorLogAndRethrow());
    }
  }

  @Override
  protected void onDetachedFromWindow() {
    if (bgSubscription != null) {
      bgSubscription.dispose();
    }
    super.onDetachedFromWindow();
  }
}

File: library/src/main/java/com/afollestad/aesthetic/AestheticImageView.kt
package com.afollestad.aesthetic

import android.content.Context
import android.support.v7.widget.AppCompatImageView
import android.util.AttributeSet
import com.afollestad.aesthetic.utils.ViewUtil
import com.afollestad.aesthetic.utils.distinctToMainThread
import com.afollestad.aesthetic.utils.onErrorLogAndRethrow
import com.afollestad.aesthetic.utils.resId
import io.reactivex.disposables.Disposable

/** @author Aidan Follestad (afollestad) */
class AestheticImageView(
  context: Context,
  attrs: AttributeSet? = null
) : AppCompatImageView(context, attrs) {

  private var bgSubscription: Disposable? = null
  private var backgroundResId: Int = 0

  init {
    if (attrs != null) {
      backgroundResId = context.resId(attrs, android.R.attr.background)
    }
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    val obs = ViewUtil.getObservableForResId(context, backgroundResId, null)
    if (obs != null) {
      bgSubscription = obs
          .distinctToMainThread()
          .subscribe(
              ViewBackgroundAction(this),
              onErrorLogAndRethrow()
          )
    }
  }

  override fun onDetachedFromWindow() {
    bgSubscription?.dispose()
    super.onDetachedFromWindow()
  }
}
